swarm/src/transport/mod.rs

```diff
@@ -108,6 +110,15 @@ pub trait Transport {
     /// doesn't recognize the protocols, or if `server` and `observed` are related.
     fn nat_traversal(&self, server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr>;
 
+    /// Applies a function on the output of the `Transport`.
+    #[inline]
+    fn map<F>(self, map: F) -> map::Map<Self, F>
+    where
+        Self: Sized,
```

Comment:
For consistency with `and_then` and `with_upgrade` should we maybe also state the constraints of  `F` here?